File: src/controllers/editable_list.ts
import { wireSubtree } from "../core/awakening";
import { observable } from "../core/decorators";
import { makeFirstResponder } from "../core/responder";
import { controllerFor } from "../core/view";
import { InlineEditor } from "./editor";
import { SortableList } from "./sortable_list";
import { Model, type ModelClass } from "../model/model";

/** List + edit lifecycle. At most one editor exists at a time (NSWindow
 *  fieldEditor pattern). The draft lives on the list as `editedObject`;
 *  the editor's `representedObject` tracks it via `bind()`. See demo.html
 *  for the `<template for="editor">` markup shape. */
export class EditableList<T> extends SortableList<T> {
  @observable accessor editedObject: T | null = null;

  private editor: InlineEditor<T> | null = null;
  private editingIndex: number | null = null;

  protected editorTemplateElement(): HTMLTemplateElement | null {
    return this.view.element.querySelector(':scope > template[for="editor"]');
  }

  // Default activation = open the editor for the selected row.
  override activateSelection(): void {
    const idx = this.selectedIndexes[0];
    if (idx == null) return;
    this.beginEditing(idx);
  }

  isEditing(): boolean {
    return this.editor !== null;
  }

  beginEditing(index: number): void {
    if (this.editor) void this.endEditing(true);

    // Editable copy that bindings can mutate without leaking into the
    // list's model until commit (Model.draft() for Models, spread otherwise).
    const original = this.arrangedObjects[index];
    let editedObject: T | null;
    if (original instanceof Model) {
      editedObject = original.draft() as T;
    } else if (original && typeof original === "object") {
      editedObject = { ...(original as object) } as T;
    } else {
      editedObject = (original ?? null) as T | null;
    }
    this.openEditor(index, editedObject);
  }

  // Mount the editor view. Used for fresh edits AND for re-entering edit
  // mode after a save failure (the same draft is preserved).
  private openEditor(index: number, editedObject: T | null): void {
    const tpl = this.editorTemplateElement();
    if (!tpl) {
      console.warn(`[Swill] ${this.constructor.name}: no <template for="editor"> inside the list view`);
      return;
    }
    const node = tpl.content.firstElementChild?.cloneNode(true) as HTMLElement | null;
    if (!node) return;

    const rowEl = this.rowElements()[index];
    if (!rowEl) return;

    this.editedObject = editedObject;

    rowEl.classList.add("being-edited");
    rowEl.after(node);
    wireSubtree(node);

    const editor = controllerFor(node) as InlineEditor<T>;
    editor.bind("representedObject", this, "editedObject");
    makeFirstResponder(editor);

    this.editor = editor;
    this.editingIndex = index;
  }

  /** Returns false only when a commit was refused (validation or save). */
  async endEditing(commit: boolean): Promise<boolean> {
    const editor = this.editor;
    const idx = this.editingIndex;
    if (!editor || idx == null) return true;

    // Validate BEFORE teardown so a rejection can keep the editor open.
    let editedObject: T | null;
    if (commit) {
      editedObject = editor.commit();
      if (editedObject instanceof Model) {
        const error = editedObject.validate();
        if (error) {
          this.editingDidFailValidation(error);
          return false;
        }
      }
    } else {
      editor.discard();
      editedObject = null;
    }

    // Destroy the editor view BEFORE the save await so it doesn't linger
    // during the round trip.
    this.editor = null;
    this.editingIndex = null;
    this.editedObject = null;
    editor._destroyView();

    if (editedObject == null) {
      const row = this.rowElements()[idx];
      row?.classList.remove("being-edited");
      makeFirstResponder(this);
      return true;
    }

    // Model drafts round-trip through save → pooled instance; plain
    // copies go in directly.
    let final: T;
    if (editedObject instanceof Model) {
      try {
        await editedObject.save();
      } catch (err) {
        // Re-mount the editor with the same draft so the user can retry.
        this.editingDidFailSave(err as Error);
        this.openEditor(idx, editedObject as T);
        return false;
      }
      const ctor = editedObject.constructor as ModelClass<Model>;
      final = (editedObject.id != null ? (ctor.store?.get(editedObject.id) ?? editedObject) : editedObject) as T;
    } else {
      final = editedObject;
    }

    const original = this.arrangedObjects[idx];
    if (original === undefined) return true;
    const sourceIndex = this.representedObject.indexOf(original);
    const arr = [...this.representedObject];
    if (sourceIndex >= 0) arr[sourceIndex] = final;
    this.representedObject = arr; // → renderAll() + clears selection
    this.selectedObject = final;
    makeFirstResponder(this);
    return true;
  }

  /** NSEditor `commitEditing:` query. Call before any action that can't
   *  proceed past in-progress edits. */
  async commitEditingIfNeeded(): Promise<boolean> {
    if (!this.editor) return true;
    return this.endEditing(true);
  }

  /** Default: prompt, validate, commit. Return false to refuse the focus
   *  transition (DOM focus snaps back to the editor). */
  protected editorShouldEndEditing(editor: InlineEditor<T>): boolean {
    const obj = editor.commit();
    if (!this.editedObjectHasChanges(obj)) {
      void this.endEditing(false);
      return true;
    }

    if (!window.confirm("Save changes?")) return false;

    if (obj instanceof Model) {
      const error = obj.validate();
      if (error) {
        this.editingDidFailValidation(error);
        return false;
      }
    }
    void this.endEditing(true);
    return true;
  }

  protected editedObjectHasChanges(obj: T | null): boolean {
    if (obj instanceof Model) return obj.isDirty;
    return obj != null;
  }

  /** Override to surface validation errors in the UI. */
  protected editingDidFailValidation(error: Error): void {
    console.warn(`[edit] ${this.constructor.name}: ${error.message}`);
  }

  /** Override to surface save errors. Editor has been re-mounted with
   *  the draft by the time this fires. */
  protected editingDidFailSave(error: Error): void {
    console.error(`[edit] ${this.constructor.name}: save failed`, error);
  }

  override cancelOperation(event: KeyboardEvent): void {
    if (this.editor) {
      void this.endEditing(false);
      return;
    }
    super.cancelOperation(event);
  }

  // The editor view lives in the rows container while editing; exclude it.
  protected override isRowElement(el: HTMLElement): boolean {
    return super.isRowElement(el) && el !== this.editor?.view.element;
  }
}

File: opal/lib/swill/controller/editable_list.rb
# backtick_javascript: true

module Swill
  class Controller::EditableList < Controller::SortableList
    property :edited_object

    def activate_selection
      index = selected_indexes.first
      index ? begin_editing(index) : super
    end

    def editing?
      !@editor.nil?
    end

    def begin_editing(index)
      return false unless index && arranged_objects[index]
      return false if editing? && !end_editing(true)

      original = arranged_objects[index]
      copy = original.respond_to?(:draft) ? original.draft : original.dup
      open_editor(index, original, copy)
    end

    def end_editing(commit)
      editor = @editor
      index = @editing_index
      original = @editing_original
      return true unless editor && index

      copy = commit ? editor.commit : nil
      if commit && (error = validation_error(copy))
        editing_did_fail_validation(error)
        return false
      end

      editor.discard unless commit
      close_editor
      if commit
        commit_edit(copy, original, index)
      else
        finish_edit(index)
      end
      true
    end

    def commit_editing_if_needed
      !editing? || end_editing(true)
    end

    def editor_should_end_editing(editor)
      copy = editor.commit
      unless edited_object_has_changes?(copy, @editing_original)
        end_editing(false)
        return true
      end
      return false unless confirm_edit?(copy, @editing_original)
      return false if (error = validation_error(copy)) && editing_did_fail_validation(error) != true

      end_editing(true)
    end

    def cancel_operation(event)
      return end_editing(false) if editing?

      super
    end

    def toggle_sort(key)
      return false unless commit_editing_if_needed

      super
    end

    protected

    def commit_edit(copy, original, index)
      final = apply_edit(copy, original, index)
      finish_edit(index, final)
      final
    end

    def apply_edit(copy, original, index)
      if original.respond_to?(:apply_draft)
        original.apply_draft(copy)
        original
      else
        source_index = represented_object.index(original)
        values = represented_object.dup
        values[source_index] = copy if source_index
        self.represented_object = values
        copy
      end
    end

    def validation_error(object)
      return unless object&.respond_to?(:validate)

      result = object.validate
      result if result.is_a?(Exception)
    rescue StandardError => error
      error
    end

    # Whether a focus-out commit should proceed. Defaults to committing;
    # applications override to interpose their own confirmation UI.
    def confirm_edit?(_copy, _original)
      true
    end

    def edited_object_has_changes?(copy, original)
      return copy.dirty? if copy.respond_to?(:dirty?)

      copy != original
    end

    def editing_did_fail_validation(error)
      warn("[Swill] #{self.class}: #{error.message}")
      false
    end

    def editing_did_fail_save(error)
      warn("[Swill] #{self.class}: save failed: #{error.message}")
    end

    def is_row_element?(element)
      super && (!@editor || `#{element} !== #{@editor.view.element}`)
    end

    # Direct child first so a nested list's editor template is not grabbed;
    # plain iteration rather than :scope>/[for=…] selectors keeps the DOM
    # shim's minimal selector engine sufficient.
    def editor_template_element
      direct = `Array.from(#{view.element}.children)`.find do |child|
        `#{child}.tagName === "TEMPLATE" && #{child}.getAttribute("for") === "editor"`
      end
      direct || `#{view.element}.querySelector('template[for="editor"]')`
    end

    def open_editor(index, original, copy)
      template = editor_template_element
      row = row_elements[index]
      unless template && row
        warn("[Swill] #{self.class}: missing <template for=\"editor\">")
        return false
      end

      node = `#{template}.content.firstElementChild && #{template}.content.firstElementChild.cloneNode(true)`
      return false unless node

      self.edited_object = copy
      `#{row}.classList.add("being-edited")`
      `#{row}.after(#{node})`
      controllers = Awakening.wire(node)
      editor = View.controller_for(node) || controllers.first
      unless editor.is_a?(Controller::InlineEditor)
        Awakening.detach(node)
        `#{node}.remove()`
        raise TypeError, "editable-list editor must use Swill::Controller::InlineEditor"
      end
      editor.bind(:represented_object, to: self, key_path: "edited_object")
      # Controls inside the editor delegate resignation to it, so a focus-out
      # commit consults editor_should_end_editing before letting go.
      editor.view.subviews.each do |subview|
        subview.delegate = editor if subview.is_a?(Control)
      end
      @editor = editor
      @editing_index = index
      @editing_original = original
      application.make_first_responder(editor)
      true
    end

    def close_editor
      editor = @editor
      @editor = nil
      @editing_index = nil
      @editing_original = nil
      self.edited_object = nil
      return unless editor

      element = editor.view.element
      Awakening.detach(element)
      `#{element}.remove()`
    end

    def finish_edit(index, final = nil)
      row = row_elements[index]
      `#{row}.classList.remove("being-edited")` if row
      self.selected_object = final if final
      application.make_first_responder(self)
    end
  end
end
